type conversion on arithmetic expression

The following program `sample.go` triggers a panic:

```go
package main

func main() {
    var p uint = 0xdead
    println(byte((1 << (p & 7)) - 1))
}   
```

Expected result:
```console
$ go run ./sample.go
31
```

Got:
```console
$ yaegi ./sample.go
./sample.go:5:15: panic
run: reflect: call of reflect.Value.SetUint on int Value
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc000360600, 0x40)
        /usr/lib/go/src/runtime/debug/stack.go:24 +0x9f
github.com/traefik/yaegi/interp.(*Interpreter).eval.func1(0xc0001cdc70)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:480 +0xc5
panic(0xe02c00, 0xc00028ef00)
        /usr/lib/go/src/runtime/panic.go:969 +0x1b9
github.com/traefik/yaegi/interp.runCfg.func1(0xc0003e40a0, 0xc0004e4a00, 0xc0001cda60)
        /home/marc/go/src/github.com/traefik/yaegi/interp/run.go:171 +0x247
panic(0xe02c00, 0xc00028ef00)
        /usr/lib/go/src/runtime/panic.go:969 +0x1b9
reflect.Value.SetUint(0xdcde00, 0xc000289be8, 0x182, 0x1f)
        /usr/lib/go/src/reflect/value.go:1705 +0x130
github.com/traefik/yaegi/interp.sub.func5(0xc0003e40a0, 0xc0004bc9f0)
        /home/marc/go/src/github.com/traefik/yaegi/interp/op.go:1033 +0x125
github.com/traefik/yaegi/interp.runCfg(0xc0004e4a00, 0xc0003e40a0)
        /home/marc/go/src/github.com/traefik/yaegi/interp/run.go:177 +0x87
github.com/traefik/yaegi/interp.(*Interpreter).run(0xc0003de000, 0xc0004e4200, 0xc0003e4000)
        /home/marc/go/src/github.com/traefik/yaegi/interp/run.go:108 +0x2af
github.com/traefik/yaegi/interp.(*Interpreter).eval(0xc0003de000, 0xc000335560, 0x56, 0x7fffac543522, 0x19, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:572 +0x809
github.com/traefik/yaegi/interp.(*Interpreter).EvalPath(0xc0003de000, 0x7fffac543522, 0x19, 0xc000335500, 0x56, 0x0, 0x0, 0xc000296d00)
        /home/marc/go/src/github.com/traefik/yaegi/interp/interp.go:394 +0xfa
main.runFile(0xc0003de000, 0x7fffac543522, 0x19, 0xc000280240, 0x1)
        /home/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:123 +0xae
main.run(0xc00000e090, 0x1, 0x1, 0x13, 0x13)
        /home/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:89 +0x7e5
main.main()
        /home/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/yaegi.go:144 +0x415
```

--- a/interp/interp_eval_test.go
+++ b/interp/interp_eval_test.go
@@ -77,6 +77,7 @@ func TestEvalShift(t *testing.T) {
 		{src: "a, b, m := uint32(1), uint32(2), uint32(0); m = a + (1 << b)", res: "5"},
 		{src: "c := uint(1); d := uint(+(-(1 << c)))", res: "18446744073709551614"},
 		{src: "e, f := uint32(0), uint32(0); f = 1 << -(e * 2)", res: "1"},
+		{src: "p := uint(0xdead); byte((1 << (p & 7)) - 1)", res: "31"},
 		{pre: func() { eval(t, i, "const k uint = 1 << 17") }, src: "int(k)", res: "131072"},
 	})
 }

--- a/interp/typecheck.go
+++ b/interp/typecheck.go
@@ -634,16 +634,13 @@ func (check typecheck) conversion(n *node, typ *itype) error {
 	if !ok {
 		return n.cfgErrorf("cannot convert expression of type %s to type %s", n.typ.id(), typ.id())
 	}
-
-	if n.typ.untyped {
-		if isInterface(typ) || c != nil && !isConstType(typ) {
-			typ = n.typ.defaultType()
-		}
-		if err := check.convertUntyped(n, typ); err != nil {
-			return err
-		}
+	if !n.typ.untyped || c == nil {
+		return nil
 	}
-	return nil
+	if isInterface(typ) || !isConstType(typ) {
+		typ = n.typ.defaultType()
+	}
+	return check.convertUntyped(n, typ)
 }
 
 type param struct {

